### ISSUE TITLE
Rivet constraint not evaluating/solving

### ISSUE THREAD

**Issue by tchesny** [david-cattermole/mayaMatchMoveSolver#176]
Hi,

I am trying to solve the translate X curve for a controller. 
When feeding mmSolver with one bundle, children of a rivet affected by this controller, the solver is not doing anything. A similar issue is happening on 0.3.8 but the result is a flat curve.
Here is a recording of a regular solve working great and the specific situation I am describing above.
I am using windows 10, maya 2018.6 and mmSolver 0.3.9

![mmSolver_issue_480](https://user-images.githubusercontent.com/32276829/97871867-5efba180-1d15-11eb-9524-c4b8c41cb074.gif)

Here's the dummy scene I am using above https://we.tl/t-iv3wIcghZr

I had a chat with Patcha yesterday about this, it works fine on his end using mmSolver 0.3.6 and maya 2018.6. It seems that the rig might be the reason and/or my maya version/install.
I am not sure to be the only on affected by this so I decided to share.

Cheers,
Thomas

**Comment by david-cattermole** [comment]
Hello Thomas,

Thank you for letting me know about this.

Do you get any information printed in the Output Window?

Thanks for providing the scene file too, that will be very helpful. I'll
have a look as soon as I can.

David

**Comment by tchesny** [comment]
Hi David,

Thanks for looking into this. 
Here is what I am getting from the Output Window :
> // Warning: mmSolver._api.execute : Invalid parameters and errors, skipping solve: [1] // 

UPDATE :
I just did the exact same test using the same maya version along with mmSolver 0.3.6 and the solve works perfectly. 
BUT it works for this controller only. other controlers like the up vector or the chest are somehow not being evaluated and produce a flat curve. Again this issue happens when using rivet to constrain bundle. Regular parent relationship are evaluated correctly.

Thomas

**Comment by david-cattermole** [comment]
Hello Thomas,

I'm starting to look at this now, and I know the cause thanks to your scene file. 
The bug is caused from a feature I added in v0.3.7 that will remove any markers or attributes from a solve that do not appear to affect the solve. You can see the status of this in the "Status" column in the Solver UI:
![image](https://user-images.githubusercontent.com/891288/97938538-c711bc80-1dd5-11eb-88e8-a98154f7e6b5.png)
A tick ✔ means "good" and a cross ✘means bad.

I will work on a fix, however if you need to solve rivets like this, I'd recommend using v0.3.6 until I can fix this.

David

**Comment by tchesny** [comment]
I noticed the little cross but was unable to change/force it to tick. Anyway, that's awesome if you can do something about it !
I will use 0.3.6 in the meantime, but keep in mind that even if it works for this specific controller on 0.3.6, I am still getting some flat curves on some controllers, like the up vector for instance. Not sure why, If I find a more particular reason I will let you right here.

The tool is still super awesome and a game changer for rotoanim as it is.

Thanks,
Thomas

**Comment by david-cattermole** [comment]
Hello @tchesny,

I believe I have fixed this problem now, and I will be releasing v0.3.10 (very soon) with the fix.

If you still have the problem please re-open this issue and I will have another look.

I'd like to thank you very much for the fantastic example scene file you've provided. 
I don't think I would have been able to fix this problem without it.

David

**Comment by david-cattermole** [comment]
Actually, it seems the new changes, although they now work correctly with character rigs and rivets now drastically increase the solving time purely DAG solves.

Here are some solve times for the different Maya versions and releases:
```
Maya 2018 v0.3.10 - avg deviation 2.24px | max deviation 13.81px at 1002 | time 991.844sec
Maya 2019 v0.3.10 - avg deviation 2.25px | max deviation 13.81px at 1002 | time 933.107sec
Maya 2018 v0.3.9 - avg deviation 2.28px | max deviation 13.81px at 1002 | time 369.991sec
Maya 2019 v0.3.9 - avg deviation 2.27px | max deviation 13.81px at 1002 | time 282.662sec
```

I will hold off releasing the new version until I can work out how to fix this performance regression.

David

**Comment by tchesny** [comment]
Hi David,

Thanks for all the details and the time you dedicated on this issue I flagged.
I am happy my maya scene helped you this much, this is a small contribution compared to the work you have done.

Any chance the fix could be pulled only when using rivet like constraints ? 

Thanks again and good luck fixing all this 😄 

Thomas

**Comment by david-cattermole** [comment]
Hello @tchesny,

Unfortunately I have not found a way to automatically 100% detect when a "complex node graph" (such as a Mesh Rivet) is present in the scene, and so I have added a new checkbox to the Standard and Basic solvers; "Evaluate Complex Node Graph (eg. Mesh Rivet)".

This is now available in v0.3.10, please install it and let me know if you still have problems.

Here are the new performance numbers for a test scene (not requiring the "Eval Complex Graphs" option):
```
Maya 2018 v0.3.10 - avg deviation 2.27px | time 334.400sec - Eval Complex Graphs=OFF
Maya 2019 v0.3.10 - avg deviation 2.27px | time 304.046sec - Eval Complex Graphs=OFF
Maya 2018 v0.3.10 - avg deviation 2.27px | time 870.519sec - Eval Complex Graphs=ON
Maya 2019 v0.3.10 - avg deviation 2.28px | time 903.377sec - Eval Complex Graphs=ON
Maya 2018 v0.3.9 - avg deviation 2.28px | time 369.991sec
Maya 2019 v0.3.9 - avg deviation 2.27px | time 282.662sec
```

I'm considering this bug is now fixed unless you find a problem

David

**Comment by tchesny** [comment]
Hi @david-cattermole,

I confirm that this bug is now fixed on my end.

Many thanks for taking care of this so quickly.

Thomas
